examples/thermostat/thermostat-common/src/atomic-write-manager.cpp

```diff
@@ -0,0 +1,535 @@
+
+    if (sessionHandle->IsSecureSession())
+    {
+        sourceNodeId = sessionHandle->AsSecureSession()->GetPeer();
+    }
+    else if (sessionHandle->IsGroupSession())
+    {
+        sourceNodeId = sessionHandle->AsIncomingGroupSession()->GetPeer();
+    }
+    return sourceNodeId;
+}
+
+/**
+ * @brief Schedules a timer for the given timeout in milliseconds.
+ *
+ * @param[in] endpoint The endpoint to use.
+ * @param[in] timeoutMilliseconds The timeout in milliseconds.
+ */
+void ThermostatAtomicWriteManager::ScheduleTimer(EndpointId endpoint, System::Clock::Milliseconds16 timeout)
+{
+    DeviceLayer::SystemLayer().StartTimer(timeout, TimerExpiredCallback,
+                                          reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+/**
+ * @brief Clears the currently scheduled timer.
+ *
+ * @param[in] endpoint The endpoint to use.
+ */
+void ThermostatAtomicWriteManager::ClearTimer(EndpointId endpoint)
+{
+    DeviceLayer::SystemLayer().CancelTimer(TimerExpiredCallback, reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+void ThermostatAtomicWriteManager::OnTimerExpired(EndpointId endpoint)
+{
+    for (size_t i = 0; i < ArraySize(mAtomicWriteSessions); ++i)
+    {
+
+        auto atomicWriteState = mAtomicWriteSessions[i];
+        if (atomicWriteState.state == kAtomicWriteState_Open && atomicWriteState.endpointId == endpoint)
+        {
+            if (mDelegate != nullptr)
+            {
+                mDelegate->OnRollbackWrite(endpoint, Presets::Id);
+                mDelegate->OnRollbackWrite(endpoint, Schedules::Id);
+            }
+            ResetWrite(endpoint);
+        }
+    }
+}
+
+void ThermostatAtomicWriteManager::SetWriteState(EndpointId endpoint, ScopedNodeId originatorNodeId, AtomicWriteState state)
+{
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        mAtomicWriteSessions[ep].state      = state;
+        mAtomicWriteSessions[ep].endpointId = endpoint;
+        mAtomicWriteSessions[ep].nodeId     = originatorNodeId;
+    }
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId, EndpointId endpoint)
+{
+    // These are the only two atomic attributes
+    if (attributeId.has_value() && attributeId.value() != Presets::Id && attributeId.value() != Schedules::Id)
+    {
+        return false;
+    }
+
+    bool inAtomicWrite = false;
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        inAtomicWrite = mAtomicWriteSessions[ep].state == kAtomicWriteState_Open;
+    }
+    return inAtomicWrite;
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId,
```

Comment:
like i commented before looks like both apis with SubjectDescriptor and commandobj check the scoped node id so maybe remove both and add an api with scopednodeid as the argument

Reply:
See below